### DISCUSSION TITLE
`pytest` randomly segfaults

### DISCUSSION THREAD

**Discussion by francesco-ballarin** [pyvista/pyvista#6295] [Q&A]
### First check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the PyVista documentation, with the integrated search.
- [X] I already searched in Google "How to X in PyVista" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I will NOT make duplicate discussions or issues in the future.

### Commit to Help

- [x] I commit to help with one of those options.

### Sample Code What is the problem, question, or error?
Write a short description telling me what you are doing, what you expect to happen, and what is currently happening.

```python
`pytest` randomly segfaults when running `pyvista` testsuite.
```

### Description

Hi,
I am trying to package `pyvista` for debian. The corresponding tracker page on debian.org is https://tracker.debian.org/pkg/python-pyvista

debian has its own CI system, on which I am trying to run `pytest` on `pyvista` testsuite using the script at https://salsa.debian.org/science-team/python-pyvista/-/blob/82f85ea25fb8f4e3b1fa93ee60f4fa43b9733dcf/debian/tests/run-unit-tests

On such CI, I am getting random segfaults, see for instance (different links correspond to different architectures)
- https://ci.debian.net/packages/p/python-pyvista/testing/arm64/48050310/
- https://ci.debian.net/packages/p/python-pyvista/testing/armhf/48050397/
- https://ci.debian.net/packages/p/python-pyvista/testing/s390x/48050551/

Has anyone of you ever had similar segfaults on headless systems?
Can you spot anything clearly wrong in how I set up the test session?
How would you suggest me to debug the failure?

Thanks,
Francesco

### System Information

```shell
--------------------------------------------------------------------------------
First block is omitted, because it would be slightly different for the 3 CI runs I linked


  Python 3.11.9 (main, Apr 10 2024, 13:16:36) [GCC 13.2.0]

           pyvista : 0.43.10
               vtk : 9.3.0
             numpy : 1.26.4
        matplotlib : 3.6.3
            scooby : 0.10.0
             pooch : 1.8.2
            pillow : 10.3.0
           imageio : 2.34.1
           IPython : 8.20.0
          colorcet : 3.0.1
             scipy : 1.12.0
              tqdm : 4.66.4
            meshio : 5.3.5
--------------------------------------------------------------------------------
```

**Comment by tkoyama010** [comment]
Thanks for letting us know that there is a Debian package for PyVista. I am excited about that.
The segmentation error you reported is also occurring in pytest, in this PyVista main repository, and we believe it comes from vtk. However, it is difficult to identify it because it occurs irregularly.

**Reply by francesco-ballarin** [reply]
Thanks @tkoyama010 !

> The segmentation error you reported is also occurring in pytest, in this PyVista main repository

Sorry, I didn't realize this was a known bug. Is there an issue I can track? Searching with `segfault` or `segmentation fault` did not help

> there is a Debian package for PyVista

The issue is that such package will _never_ be considered for inclusion in a Debian release unless I can get CI to go green, and to reliably keep running without random failures when testing future updates of PyVista and its dependencies.
What workaround do you use in your CI? For instance, I'd be open to skip some tests if you know from experience which ones are causing the problem.

**Reply by user27182** [reply]
> Is there an issue I can track? Searching with segfault or segmentation fault did not help

There are a few open issues tracking flaky tests generally, which include some caused by segmentation faults, e.g. https://github.com/pyvista/pyvista/issues/5334 and https://github.com/pyvista/pyvista/issues/3054.

> What workaround do you use in your CI?

I believe we currently just re-try failed CI tests manually. There is a [discussion](https://github.com/pyvista/pyvista/issues/5334#issuecomment-1853482215) to possibly use `pytest-repeat` though as a workaround. Or, as you suggest, you could try skipping them altogether.

**Reply by MatthewFlamm** [reply]
Can you explain how the packaging is done for the tests, particularly for vtk?  I'm seeing this line in the logs:

> required packages: vtk (not found)

We test, for the most part, against publicly released wheels of vtk.  It seems like the testing is using some different version of vtk.  Assuming that debian has a custom package of vtk, it isn't too surprising.

Can you run pytest in verbose mode to see if it the same test over and over and it could be marked either here or there in some way?

**Reply by francesco-ballarin** [reply]
Probably the most informative file about `vtk` packaging is https://salsa.debian.org/science-team/vtk9/-/blob/7f2776268474cc9cba80a34a39a0b3c6b3d4920e/debian/rules . I am not involved (yet?) in that package, so I don't know much details. Looking at https://github.com/pyvista/pyvista/blob/ed4f31184749585dc8651d87a368bc4b2a3a3879/tests/conftest.py#L250 , I'd guess that debian's `vtk9` package is missing the relevant `metadata`.

I am can surely increase verbosity and make another upload of the `pyvista` package. I'll come back with new links when CI runs them.

**Reply by francesco-ballarin** [reply]
I've applied the suggestions we discussed. It will take me longer than expected to share the updated CI runs, because Debian (unstable) transitioned yesterday to python 3.12, and now I'll have to wait for all `pyvista` dependencies (most notably, `vtk`) to be rebuilt with 3.12 instead of 3.11.

**Comment by francesco-ballarin** [comment]
OK, so as promised I am back with some test results.

The updated test script is at https://salsa.debian.org/science-team/python-pyvista/-/blob/ce2a7370f38ea6e270326d15ec4acf7ac99bfaf4/debian/tests/run-unit-tests , and from line 16 to line 110 I list some tests I already skip, and the corresponding motivation.

Please note that on line 123 I add `--random-order` to `pytest` options to randomize the order in which tests are run, which has helped me debugging if segmentation fault on the same architechture is always caused by the same test, or if there are multiple tests that cause it.

---

**The good news**: there are three architectures with 100% success rate on CI (i.e, no segmentation fault on repeated runs), including the widely used amd64

amd64: 100% success
- https://ci.debian.net/packages/p/python-pyvista/unstable/amd64/48366652/
- https://ci.debian.net/packages/p/python-pyvista/unstable/amd64/48355607/
- https://ci.debian.net/packages/p/python-pyvista/unstable/amd64/48165193/
- https://ci.debian.net/packages/p/python-pyvista/unstable/amd64/48376091/

armel (32 bit arch): 100% success
- https://ci.debian.net/packages/p/python-pyvista/unstable/armel/48366654/
- https://ci.debian.net/packages/p/python-pyvista/unstable/armel/48376093/

ppc64el: 100% success
- https://ci.debian.net/packages/p/python-pyvista/unstable/ppc64el/48366658/
- https://ci.debian.net/packages/p/python-pyvista/unstable/ppc64el/48376097/

---

**The bad news, part 1**: there are two archs in which at least one test is randomly segfaulting (i.e., segfaulting in one run, but not in another)

arm64: 75% success
*Failing*
- https://ci.debian.net/packages/p/python-pyvista/unstable/arm64/48355486/
```
tests/plotting/test_plotting.py::test_plot_above_below_color Fatal Python error: Aborted
```
*Successful*
- https://ci.debian.net/packages/p/python-pyvista/unstable/arm64/48355485/
- https://ci.debian.net/packages/p/python-pyvista/unstable/arm64/48366653/
- https://ci.debian.net/packages/p/python-pyvista/unstable/arm64/48376092/

i386 (32 bit arch): 50% success
*Failing*
- https://ci.debian.net/packages/p/python-pyvista/unstable/i386/48376095/
```
tests/core/test_reader.py::test_hdr_reader Fatal Python error: Aborted
```
*Successful* (technically it failed, but the failure here is not a segfault)
- https://ci.debian.net/packages/p/python-pyvista/unstable/i386/48366656/
```
tests/core/test_dataset.py::test_rotate_should_match_vtk_rotation[z]
tests/core/test_dataset.py::test_rotate_should_match_vtk_rotation[y]
-> I'll simply skip these tests
```

---

**The bad news, part 2**: there are two archs in which I've never been able to get a successful run 

s390x: 0% success
- https://ci.debian.net/packages/p/python-pyvista/unstable/s390x/48366660/
- https://ci.debian.net/packages/p/python-pyvista/unstable/s390x/48315315/
- https://ci.debian.net/packages/p/python-pyvista/unstable/s390x/48376099/
```
segfault at the very beginning, before executing any test
```

armhf (32 bit arch): 0% success
- https://ci.debian.net/packages/p/python-pyvista/unstable/armhf/48366655/
```
tests/plotting/test_widgets.py::test_clear_logo_widget Fatal Python error: Segmentation fault
```
- https://ci.debian.net/packages/p/python-pyvista/unstable/armhf/48376094/
```
tests/plotting/test_picking.py::test_surface_point_picking[True] Fatal Python error: Segmentation fault
```

---

**The bad news, part 3**: in this list, the same segfault never appeared twice (!). Of course I can add them tests to the blacklist as soon as I see them failing once, but maintaining the blacklist may be tedious and error prone in future. Can you see any pattern forming in the list of tests which segfaulted up to now?
